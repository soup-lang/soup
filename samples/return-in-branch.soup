func aReturn(x, cb)
    setTimeout(%, 1)
    if x
        cb(null, x - 1)
    else
        cb('a', null)

func sReturn(x)
    if x
        return x - 1
    else
        return null

func ss(a, b, x)
    if a
        if b
            return sReturn(x + 2)
        return sReturn(x + 1)
    if b
        return sReturn(x)
    return sReturn(x - 1)

func sa(a, b, x, %%)
    if a
        if b
            setTimeout(%, 1)
            return aReturn(x + 2, %%)
        return sReturn(x + 1)
    if b
        setTimeout(%, 1)
        return aReturn(x, %%)
    return sReturn(x - 1)

func as(a, b, x, %%)
    if a
        setTimeout(%, 1)
        if b
            return sReturn(x + 2)
        return aReturn(x + 1, %%)
    if b
        return sReturn(x)
    return sReturn(x - 1)

func aa(a, b, x, %%)
    setTimeout(%, 1)
    if a
        setTimeout(%, 1)
        if b
            setTimeout(%, 1)
            return aReturn(x + 2, %%)
        return aReturn(x + 1, %%)
    if b
        return aReturn(x, %%)
    return aReturn(x - 1, %%)

console.log('ss')
console.log(ss(true, true, 3))
console.log(ss(true, false, 3))
console.log(ss(false, true, 3))
console.log(ss(false, false, 3))

console.log('sa')
console.log(sa(true, true, 13, %%))
console.log(sa(true, false, 13, %%))
console.log(sa(false, true, 13, %%))
console.log(sa(false, false, 13, %%))

console.log('as')
console.log(as(true, true, 23, %%))
console.log(as(true, false, 23, %%))
console.log(as(false, true, 23, %%))
console.log(as(false, false, 23, %%))

console.log('aa')
console.log(aa(true, true, 33, %%))
console.log(aa(true, false, 33, %%))
console.log(aa(false, true, 33, %%))
console.log(aa(false, false, 33, %%))
