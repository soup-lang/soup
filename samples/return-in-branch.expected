ss
4
3
2
1
sa
14
13
12
11
as
24
23
22
21
aa
34
33
32
31
